sensors/gps/nmea.c

```diff
@@ -0,0 +1,313 @@
+		return nmea_broken;
+	}
+	/* when error occurred 0 is assigned */
+	in.h_wgs = strtod(p + 1, NULL);
+
+	out->msg.gga = in;
+	out->type = nmea_gga;
+	return nmea_gga;
+}
+
+
+static int nmea_parsermc(char *str, nmea_t *out)
+{
+	char *p = str, *sign;
+	nmea_rmc_t in;
+
+	/* latitude read */
+	if (nmea_nField(&p, p, field_rmc_lat) == NULL || nmea_nField(&sign, p, 1) == NULL) {
+		return nmea_broken;
+	}
+	/* when error occurred 0 is assigned */
+	in.lat = strtod(p + 1, NULL);
+	in.lat = ((int)in.lat / 100) + (in.lat - ((int)in.lat / 100) * 100) / 60; /* conversion from ddmm.mmmm to dd.dddd */
+	in.lat = (sign[1] == 'S') ? -in.lat : in.lat;
+
+	/* longitude read */
+	if (nmea_nField(&p, p, field_rmc_lon - field_rmc_lat) == NULL || nmea_nField(&sign, p, 1) == NULL) {
+		return nmea_broken;
+	}
+	/* when error occurred 0 is assigned */
+	in.lon = strtod(p + 1, NULL);
+	in.lon = ((int)in.lon / 100) + (in.lon - ((int)in.lon / 100) * 100) / 60; /* conversion from ddmm.mmmm to dd.dddd */
+	in.lon = (sign[1] == 'W') ? -in.lon : in.lon;
+
+	/* speed in knots read */
+	if (nmea_nField(&p, p, field_rmc_speedknots - field_rmc_lon) == NULL) {
+		return nmea_broken;
+	}
+	/* when error occurred 0 is assigned */
+	in.speed = strtod(p + 1, NULL);
+
+	/* course read */
+	if (nmea_nField(&p, p, field_rmc_course - field_rmc_speedknots) == NULL) {
+		return nmea_broken;
+	}
+	/* when error occurred 0 is assigned */
+	in.course = strtod(p + 1, NULL);
+
+	/* magnetic variation read */
+	if (nmea_nField(&p, p, field_rmc_magvar - field_rmc_course) == NULL) {
+		return nmea_broken;
+	}
+	/* when error occurred 0 is assigned */
+	in.magvar = strtod(p + 1, NULL);
+
+	out->msg.rmc = in;
+	out->type = nmea_rmc;
+	return nmea_rmc;
+}
+
+
+int nmea_interpreter(char *str, nmea_t *out)
+{
+	int ret;
+
+	if (strncmp(str, "$GPGSA", 6) == 0) {
+		ret = nmea_parsegsa(str, out);
+	}
+	else if (strncmp(str, "$GPVTG", 6) == 0) {
+		ret = nmea_parsevtg(str, out);
+	}
+	else if (strncmp(str, "$GPGGA", 6) == 0) {
+		ret = nmea_parsegga(str, out);
+	}
+	else if (strncmp(str, "$GPRMC", 6) == 0) {
+		ret = nmea_parsermc(str, out);
+	}
+	else {
+		ret = nmea_unknown;
+	}
+
+	return ret;
+}
+
+
+int nmea_countlines(char *buf)
+{
+	int n = 0;
+	char *start, *curr;
+
+	start = strchr(buf, '$');
+	curr = start;
+
+	while (curr != NULL) {
+		/* assert that '*' and checksum characters are available */
+		curr = strchr(curr, '*');
+		if (curr != NULL) {
+			if ((curr[1] != '\0') && (curr[2] != '\0')) {
+				n++;
+			}
+			curr = strchr(curr, '$');
+		}
+	}
+
+	return n;
+}
+
+
+char **nmea_getlines(char *buf, int n)
+{
+	char **lines;
+	char *curr, *start;
+	int i;
+	curr = buf;
+
+	lines = malloc(sizeof(char *) * n);
+
+	if (lines != NULL) {
+		for (i = 0; i < n; i++) {
+			start = strchr(curr, '$');
+			if (start != NULL) {
+				curr = strchr(start, '*');
+				if (curr != NULL) {
+					curr[3] = '\0';
+					curr += 4;
+				}
+				lines[i] = start;
+			}
```

Comment:
I was assuming that input buffer is appropriately prepared, but yeah it could be done in a better way. Here is my proposal:
```C
char **nmea_getlines(char *buf, int n)
{
	char **lines;
	char *curr, *start;
	int i;

	curr = buf;

	lines = calloc(n, sizeof(char *));

	if (lines == NULL) {
		return NULL;
	}

	for (i = 0; i < n; i++) {
		start = strchr(curr, '$');
		if (start == NULL) {
			break;
		}

		curr = strchr(start, '*');
		if (curr == NULL) {
			break;
		}

		curr[3] = '\0';
		curr += 4;

		lines[i] = start;
	}

	return lines;
}
```